homeassistant/components/improv_ble/config_flow.py

```diff
@@ -125,22 +115,66 @@ async def async_step_user(
             errors=errors,
         )
 
-    async def async_step_bluetooth(
-        self, discovery_info: BluetoothServiceInfoBleak
-    ) -> FlowResult:
-        """Handle the Bluetooth discovery step."""
-        await self.async_set_unique_id(discovery_info.address)
-        self._abort_if_unique_id_configured()
-        service_data = discovery_info.service_data
+    def _abort_if_provisioned(self) -> None:
+        """Check improv state and abort flow if needed."""
+        # mypy is not aware that we can't get here without having these set already
+        assert self._discovery_info is not None
+
+        service_data = self._discovery_info.service_data
         improv_service_data = ImprovServiceData.from_bytes(
             service_data[SERVICE_DATA_UUID]
         )
         if improv_service_data.state in (State.PROVISIONING, State.PROVISIONED):
             _LOGGER.debug(
-                "Device is already provisioned: %s", improv_service_data.state
+                "Aborting improv flow, device is already provisioned: %s",
+                improv_service_data.state,
             )
-            return self.async_abort(reason="already_provisioned")
+            raise AbortFlow("already_provisioned")
+
+    @callback
+    def _async_update_ble(
+        self,
+        service_info: bluetooth.BluetoothServiceInfoBleak,
+        change: bluetooth.BluetoothChange,
+    ) -> None:
+        """Update from a ble callback."""
+        _LOGGER.debug(
+            "Got updated BLE data: %s",
+            service_info.service_data[SERVICE_DATA_UUID].hex(),
+        )
+
+        self._discovery_info = service_info
+        try:
+            self._abort_if_provisioned()
+        except AbortFlow:
+            self.hass.config_entries.flow.async_abort(self.flow_id)
+
+    def _unregister_bluetooth_callback(self) -> None:
+        """Unregister bluetooth callbacks."""
+        if not self._remove_bluetooth_callback:
+            return
+        self._remove_bluetooth_callback()
+        self._remove_bluetooth_callback = None
+
+    async def async_step_bluetooth(
+        self, discovery_info: bluetooth.BluetoothServiceInfoBleak
+    ) -> FlowResult:
+        """Handle the Bluetooth discovery step."""
         self._discovery_info = discovery_info
+
+        await self.async_set_unique_id(discovery_info.address)
+        self._abort_if_unique_id_configured()
+        self._abort_if_provisioned()
+
+        self._remove_bluetooth_callback = bluetooth.async_register_callback(
+            self.hass,
+            self._async_update_ble,
+            bluetooth.BluetoothCallbackMatcher(
+                {bluetooth.match.ADDRESS: discovery_info.address}
+            ),
+            bluetooth.BluetoothScanningMode.PASSIVE,
+        )
```

Comment:
I'm having trouble following why we register for callbacks and than move to `self.async_step_bluetooth_confirm` and than remove the callback subscription right away by calling  `self._unregister_bluetooth_callback()`?

Was the intent to do that after `if user_input is None:`  ?